### PR TITLE
[17009] Keep ID of inlined element while insertText is used. 

### LINKED ISSUE #697: insertText splits spans with id attribute into two
## Are you reporting a feature or a bug?

Bug

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

https://dev.ckeditor.com/ticket/17009

## Provide detailed reproduction steps (if any)

1. Prepare CKEditor with below configuration and content:
	```js
	var editor = CKEDITOR.replace( 'editor1', {
		extraAllowedContent : 'span[id]',
	});
	```

	```html
	<p><span class="marker" id="abc">test</span></p>
	```
2. Put selection here and insert any special character ```#``` using Special Character dialog ```<p><span class="marker" id="abc">te^st</span></p>``` . Alternatively you can ma a selection and execute the following command in browser console ```CKEDITOR.insatnces.editor1.insertText( '#' )```

### Expected result

```<p><span class="marker" id="abc">te#st</span></p>```

### Actual result

The span gets split in two: ```<p><span class="marker" id="abc">te</span><span class="marker">#st</span></p>```

## Other details

Problem can be reproduced at least from CKEditor 4.0.

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?
Bug fix

## Does your PR contain necessary tests?
yes

### This PR contains
- [x] Unit tests
- [x] Manual tests

## What changes did you make?
- Change in method `mergeSibling` to always combine elements with higher index to element with lower index.
- Create mechanism to preserve `id`:
-- firstly `id` with reference to element is kept in array
-- because merging siblings was change, I have guarantee that reference to element will remain for entire process
-- then `id` is removed from elements
-- now process of `insert` is going as usual. If there will be possibility to merge inline element, this should happen
-- `id` is restored to elements. If there will be inserted non-inline elements, then `id` will apply to first element.

### Before change:
- `<p><span id="abc">te^st</span></p>` + `insertText('FOO');`
- `<p><span id="abc">te</span><span>FOOst</span></p>`

### After change:
- `<p><span id="abc">te^st</span></p>` + `insertText('FOO');`
- `<p><span id="abc">teFOOst</span></p>`

Closes #697.

### CODE REVIEW

File: core/dom/element.js [hunk 1]
```diff
@@ -1165,20 +1165,30 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
...
 					// queuing them to be moved later. (#5567)
-					var pendingNodes = [];
 
+					var pendingNodes = [],
+						temporaryElement, node;
 					while ( sibling.data( 'cke-bookmark' ) || sibling.isEmptyInlineRemoveable() ) {
 						pendingNodes.push( sibling );
 						sibling = isNext ? sibling.getNext() : sibling.getPrevious();
-						if ( !sibling || sibling.type != CKEDITOR.NODE_ELEMENT )
+						if ( !sibling || sibling.type != CKEDITOR.NODE_ELEMENT ) {
```
Comment (msamsel) [inline core/dom/element.js]: I add parenthesis in missing places in this function.

File: core/dom/element.js [hunk 2]
```diff
@@ -1165,20 +1165,30 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
...
 						// Save the last child to be checked too, to merge things like
 						// <b><i></i></b><b><i></i></b> => <b><i></i></b>
 						var innerSibling = isNext ? element.getLast() : element.getFirst();
 
+						// #17009 swap elements, to analyse then in this same order (from lower index to higher)
+						if ( element.getIndex() > sibling.getIndex() ) {
+							temporaryElement = element;
+							element = sibling;
+							sibling = temporaryElement;
+							isNext = !isNext;
```
Comment (msamsel) [inline core/dom/element.js]: Test is added in `tests/core/dom/element/element.js`

File: tests/core/editable/manual/insertionkeepid.md [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
-1. Put carret in editor.
-1. Insert text, by usinng `Insert Special Character` button.
-1. Check `Source` again
-1. There should remain one `spane` element with `id`
-1. Repeat steps in 2nd editor
-
-**Unexpected:** As result you will obtain splited text into 2 span elements.
+1. Switch to `Source mode` and make sure there is a `span` with `id` attribute.
+1. Switch back to `WYSIWYG mode` and put caret inside `Foo` word.
+1. Insert character usinng `Insert Special Character` button.
```
Comment (f1ames) [inline tests/core/editable/manual/insertionkeepid.md]: > character **using** ...

File: tests/core/editable/manual/insertionkeepid.md [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
+1. Check `Source mode` again.
+
+**Expected:** There should remain one `span` element with `id`.
+
+**Unexpected:** After specal character insertion `span` element was splited into two: e.g. `<span id="SomeId2">Fo#</span><span>o</span>`.
```
Comment (f1ames) [inline tests/core/editable/manual/insertionkeepid.md]: > element was splited into

Past tense is **split**.

File: tests/core/editable/insertionpreserveid.js [hunk 5]
```diff
@@ -1,41 +1,40 @@
...
 		}
 	};
 
-	bender.test({
+	bender.test( {
 		'test insert nothing into inline element with id': function() {
 			var editor = this.editors.paragraph,
 				bot = this.editorBots.paragraph;
-			bot.setHtmlWithSelection( '<p style="text-align: right;"><span class="marker" id="abc">te^st</span></p>' );
+			bot.setHtmlWithSelection( '<p style="text-align: right"><span class="marker" id="abc">te^st</span></p>' );
```
Comment (f1ames) [inline tests/core/editable/insertionpreserveid.js]: The `bot.setHtmlWithSelection` uses deprecated `bender.tools.setHtmlWithSelection`, you should use `bender.tools.selection.setWithHtml` instead.

File: core/dom/element.js [hunk 6]
```diff
@@ -1186,7 +1180,7 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
 						// <b><i></i></b><b><i></i></b> => <b><i></i></b>
 						var innerSibling = isNext ? element.getLast() : element.getFirst();
 
-						// #17009 swap elements, to analyse then in this same order (from lower index to higher)
+						// Swap elements to return always this same order: from lowest to highest index (https://dev.ckeditor.com/ticket/17009).
```
Comment (f1ames) [inline core/dom/element.js]: > always **the** same

or

> to return **them always in the** same

File: core/editable.js [hunk 7]
```diff
@@ -1716,17 +1716,18 @@
...
-				// #17009 store elements with id. Beacuse siblings are marged always from lower child index to higher, refrence to node will persist after all operations.
-				for ( i=0; i < path.elements.length; i++ ) {
-					element = path.elements[i];
-					if ( element.hasAttribute( 'id' ) ){
-						that.elementsWithId.push({
-							"node": element,
-							"id": element.getAttribute('id')
-						});
-						element.removeAttribute('id');
+				// Store elements with id, beacuse siblings are marged always from lowest child index to highest,
```
Comment (f1ames) [inline core/editable.js]: > are **merged** always

File: core/editable.js [hunk 8]
```diff
@@ -1967,6 +1980,11 @@
 
 			range.moveToBookmark( bm );
 
+			// #17009 restore id to previously stored elements.
```
Comment (f1ames) [inline core/editable.js]: Remember about proper ticket referencing (full Trac url on the end).

**Comment by msamsel** [comment]
Note:
Currently realisation is paused.
Issue requires few decision to be made related to ID behaviour.
`<0>` - element without ID
`<IDx>` - element with some ID
`^` - cursor
- <ID1>Some^ text</ID1> + <0>input</0> -> what should be done with right part of ID1 element. What if there will be some classes or inline styles?
- <ID1>Some^ text</ID1> + <ID2>input</ID2> -> Similar question as above.
- <ID1 class="some fancy class">Some^ text</ID1> + <ID1>input</ID1> -> what if ID match but styles not?

Problem exists because `insertText` and `insertHtml` use this same method for inserting text.

**Pull request closed without merging**